### DISCUSSION TITLE
Concatenating MIDI files

### DISCUSSION THREAD

**Discussion by frallebini** [mido/mido#444] [Q&A] [answered]
Given two (type 1) MIDI files
```python
mid1 = MidiFile('first.mid')
mid2 = MidiFile('second.mid')
```
how could I go about creating a
```python
concat_mid = MidiFile()
```
such that `concat_mid` "plays" `mid1`'s tracks first, and then, when `mid1` is over, it starts playing `mid2`'s tracks?

**Comment by frallebini** [comment, marked as answer]
I think I found a solution. I'll share it here for future reference:
```python
mid1 = MidiFile('first.mid')
mid2 = MidiFile('second.mid')

# compute mid1 total time
time1 = 0
for track in mid1.tracks:
    for message in track:
        if message.type == 'note_on':
            time1 += message.time

# shift mid2 by time1
for track in mid2.tracks:
    for message in track:
        if message.type == 'program_change':
            message.time += time1

# merge mid1 and shifted mid2
concat_mid = MidiFile()
concat_mid.tracks = mid1.tracks + mid2.tracks
concat_mid.save('concat.mid')
```

**Comment by rdoursenaud** [comment]
This would have been better suited for our [Discussions](https://github.com/mido/mido/discussions) area.
Note that your solution is only valid for two files using the same resolution.

**Comment by frallebini** [comment]
Sorry, didn’t realize there was such a thing.

What do you mean by “resolution”?

**Comment by rdoursenaud** [comment]
Every Standard Midi File (SMF) has a "ticks per quarter note" information in the header. The delta-time used by messages is defined in ticks. Hence the resolution bit.

*But* I've made a mistake stating it wouldn't work since `mido` seems to be encoding the `time` information in seconds so my bit is irrelevant.
Sorry for the confusion.
